whittle/lora.py

```diff
@@ -0,0 +1,821 @@
+        sampler: Sampler-related arguments. See ``whittle.args.SamplerArgs`` for details.
+    """
+
+    checkpoint_dir = auto_download_checkpoint(
+        model_name=checkpoint_dir, access_token=access_token
+    )
+    pprint(locals())
+    data_str = dataset
+    if data_str == "alpaca":
+        data = Alpaca()
+    elif data_str == "llamamini":
+        data = LLaMaMini()
+    else:
+        data = None
+    devices = parse_devices(devices)
+    out_dir = init_out_dir(out_dir)
+
+    check_valid_checkpoint_dir(checkpoint_dir)
+    config = Config.from_file(
+        checkpoint_dir / "model_config.yaml",
+        lora_r=lora_r,
+        lora_alpha=lora_alpha,
+        lora_dropout=lora_dropout,
+        lora_query=lora_query,
+        lora_key=lora_key,
+        lora_value=lora_value,
+        lora_projection=lora_projection,
+        lora_mlp=lora_mlp,
+        lora_head=lora_head,
+        lora_emb=lora_emb,
+    )
+    config.fix_head_size = True
+    config.tie_embeddings = False
+    config.model_type = "gpt"
+    now = datetime.now()
+    sampling_strategy = sampler.sampling_strategy
+    seed_sampler = sampler.seed_sampler
+    num_configs = sampler.num_configs
+    n_trials = sampler.n_trials
+    # Create a timestamp with nanosecond precision
+    time_string = now.strftime("%Y%m%d_%H%M%S")
+    search_space = search_spaces[search_space_type](config)
+    out_dir = Path(
+        f"{config.name}-{train_strategy}-{search_space_type}-{sampling_strategy}-{data_str}/finetune/lora/"
+    )
+    id = f"{train_strategy}-{search_space_type}-{sampling_strategy}-{time_string}-{data_str}-lora"
+    precision = precision or get_default_supported_precision(training=True)
+    logger = choose_logger(
+        logger_name,
+        out_dir,
+        name=f"finetune-{config.name}",
+        log_interval=train.log_interval,
+        id=id,
+        resume=bool(resume),
+        config=dict(
+            train_strategy=train_strategy,
+            search_space_type=search_space_type,
+            sampling_strategy=sampling_strategy,
+            data=data_str,
+            lora_r=lora_r,
+            lora_alpha=lora_alpha,
+            lora_emb=lora_emb,
+            lora_mlp=lora_mlp,
+            lora_head=lora_head,
+            lora_projection=lora_projection,
+            lr_warmup_steps=train.lr_warmup_steps,
+        ),
+    )
+
+    plugins = None
+
+    if devices * num_nodes > 1:
+        strategy = DDPStrategy(find_unused_parameters=True)
+    else:
+        strategy = "auto"
+
+    fabric = L.Fabric(
+        devices=devices,
+        num_nodes=num_nodes,
+        strategy=strategy,
+        precision=precision,
+        loggers=logger,
+        plugins=plugins,
+    )
+
+    sampler = get_sampler(
+        sampling_strategy,
+        search_space=search_space,
+        seed=seed_sampler,
+        num_configs=num_configs,
+        n_trials=n_trials,
+    )
+    if train_strategy == "sandwich":
```

Comment:
Could we re-use  `training_strategies_cls` which would make it easier to add new training strategies in the future?